### ISSUE TITLE
'A' auto inserted when Telescope is run from insert mode

### ISSUE THREAD

**Issue by thefallenidealist** [nvim-telescope/telescope.nvim#2195]
### Description

char 'A' is automatically inserted into Telescope prompt if Telescope is called from insert mode with <C-o>. Same with innoremap.

### Neovim version

```markdown
NVIM v0.8.0-v0.8.0                                                                                                                                                                            
Build type: RelWithDebInfo                                                                                                                                                                    
LuaJIT 2.1.0-beta3
```

### Operating system and version

FreeBSD 13.1-RELEASE-p2

### Telescope version / branch / rev

master f174a03

### checkhealth telescope

```markdown
telescope: require("telescope.health").check()
========================================================================
## Checking for required plugins
  - OK: plenary installed.
  - OK: nvim-treesitter installed.

## Checking external dependencies
  - OK: rg: found ripgrep 13.0.0
  - WARNING: fd: not found. Install [sharkdp/fd](https://github.com/sharkdp/fd) for extended capabilities

## ===== Installed extensions =====

## Telescope Extension: `blines`
  - INFO: No healthcheck provided

## Telescope Extension: `file_browser`
  - INFO: No healthcheck provided

## Telescope Extension: `live_grep_args`
  - INFO: No healthcheck provided
```

### Steps to reproduce

open nvim
goto insert mode
Press <C-o> and enter ":Telescope"

### Expected behavior

'A' not inserted in Telescope prompt

### Actual behavior

'A'  inserted in Telescope prompt

### Minimal config

```Lua
local Plug = vim.fn['plug#']
vim.call('plug#begin', '~/.config/nvim/plugged')

Plug 'nvim-lua/popup.nvim'
Plug 'nvim-lua/plenary.nvim'
Plug 'nvim-telescope/telescope.nvim'
vim.call('plug#end')

require('telescope').setup{}
```

**Comment by thefallenidealist** [comment]
And this patch which I forgot fixes the issue:
```
diff --git a/lua/telescope/pickers.lua b/lua/telescope/pickers.lua
index 5019abf..f400c80 100644
--- a/lua/telescope/pickers.lua
+++ b/lua/telescope/pickers.lua
@@ -430,14 +430,15 @@ function Picker:find()
   end
 
   if vim.tbl_contains({ "insert", "normal" }, self.initial_mode) then
-    local mode = vim.fn.mode()
+    -- local mode = vim.fn.mode()        -- after <C-o> returns 'n'
+    local mode = vim.api.nvim_get_mode().mode    -- after <C-o> returns 'niI'
     local keys
     if self.initial_mode == "normal" then
       -- n: A<ESC> makes sure cursor is at always at end of prompt w/o default_text
       keys = mode ~= "n" and "<ESC>A<ESC>" or "A<ESC>"
     else
       -- always fully retrigger insert mode: required for going from one picker to next
-      keys = mode ~= "n" and "<ESC>A" or "A"
+      keys = mode ~= "n" and "<ESC>A" or "A"     -- XXX
     end
     a.nvim_feedkeys(a.nvim_replace_termcodes(keys, true, false, true), "n", true)
   else
```

**Comment by fdschmidt93** [comment]
Duplicate of https://github.com/nvim-telescope/telescope.nvim/issues/2052 ?

tl;dr: <C-o> is not needed, but looking at your patch `vim.api.nvim_get_mode().mode` might fix it in general? Why not make a PR and then ideally check that the table at bottom of opening post in https://github.com/nvim-telescope/telescope.nvim/pull/1917#issue-1225171325 still works. This table is also essentially the reason why I'd be happier if solution of above linked issue would just be fine :)

**Comment by thefallenidealist** [comment]
Yes, it looks like duplicate of #2052 
Tested it again on latest nvim master and latest Telescope - problem still reproducible.

However, my original problem was when I use this shortcut (blines is local extension, imap() is wrapper around vim.keymap.set()):
`imap('<C-x><C-l>', '<C-o>:Telescope blines<CR>')`

Problem is not reproducible with:
`imap('<C-x><C-l>', '<cmd>:Telescope blines<CR>')`

**Comment by reaink** [comment]
Also have this problem

**Comment by jamestrew** [comment]
@reaink are you using `<cmd>:Telescope ...` to do your mapping. As stated earlier, opening telescope from insert mode with this method seems to work.

I don't recommend using `<C-o>:Telescope ...`. `<C-o>` seems to have a few documented side-effects (not sure if they fully relate to this issue) and the docs for `:h <Cmd>` also states
> This is more flexible than `:<C-U>` in Visual and Operator-pending mode, or `<C-O>:` in Insert mode, because the commands are executed directly in the current mode, instead of always going to Normal mode.

I'm going to closed this issue for these reasons. If you're using `<cmd>:` and still having issue, we can reopen and discuss further.
